### PR TITLE
Return of hook profiling

### PR DESCRIPTION
**Pull request by StanislovasRamanauskas**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | Added hook profiling which was removed in 1.7.0.0 version.
| Type?         | improvement
| Category?     | FO / BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/9659
| How to test?  | Turn on _PS_DEBUG_PROFILING_ in defines.inc.php and see profiling list in page bottom.<br>Screenshot from 2019-10-31 16-29-39

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16231)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hi @StanislovasRamanauskas, why not use the Hook profiler provided in the current debug toolbar ?

**Comment by kpodemski** [comment]
> Hi @StanislovasRamanauskas, why not use the Hook profiler provided in the current debug toolbar ?

because you don't have symfony profiler on front-office @mickaelandrieu :)

**Comment by mickaelandrieu** [comment]
> because you don't have symfony profiler on front-office @mickaelandrieu :)

I always forget the front office !

File: install-dev/upgrade/sql/1.7.0.0.sql [hunk 1]
```diff
@@ -37,8 +37,6 @@ ALTER TABLE `PREFIX_customer` CHANGE COLUMN `lastname` `lastname` varchar(255) N
 /* Changes regarding modules */
 ALTER TABLE `PREFIX_module` ADD UNIQUE INDEX `name_UNIQUE` (`name` ASC);
 
-DROP TABLE IF EXISTS `PREFIX_modules_perfs`;
```
Comment (sarjon) [inline install-dev/upgrade/sql/1.7.0.0.sql]: Table was dropped in 1.7.0.0, you don't need to remove it from here, but create it once again in `1.7.7.0.sql` file.

**Comment by StanislovasRamanauskas** [comment]
> Hi @StanislovasRamanauskas, why not use the Hook profiler provided in the current debug toolbar ?
@mickaelandrieu hook debugging is important feature for improving module speed. Also, not all hook profiling functionality was removed from 1.7.0.0 version. Till now PS have empty hook list in profiling and that is confusing. 
Screenshot from 2019-11-03 10-35-10

**Comment by kpodemski** [comment]
@PrestaShop/prestashop-core-developers LGTM!

File: classes/Hook.php [hunk 2]
```diff
@@ -967,7 +967,44 @@ public static function exec(
 
     public static function coreCallHook($module, $method, $params)
     {
-        return $module->{$method}($params);
+        // Define if we will log modules performances for this session
+        if (Module::$_log_modules_perfs === null) {
+            Module::$_log_modules_perfs = (_PS_DEBUG_PROFILING_ && mt_rand(0, _PS_DEBUG_PROFILING_ - 1) == 0);
+            if (Module::$_log_modules_perfs) {
+                Module::$_log_modules_perfs_session = mt_rand();
```
Comment (eternoendless) [inline classes/Hook.php:974]: Looking at the code, it looks that the session ID is random and lasts for exactly one run. In that case, why bother to save results on the database?

**Review by eternoendless — CHANGES_REQUESTED** [review]
I think this could be useful for some uses, but I think the implementation needs to be improved. For example, I'd rather see the whole profiling logic in a class of its own:

```php
function doSomething(...$params) {
  if ($this->profiler->shouldProfile()) {
    return $this->profiler->interceptDoSomething(...$params); // does the thing with profiling
  }

  return $this->doTheThing(...$params); // does the thing without profiling
}
```

or even better, use strategy pattern:

```php
public function doSomething(...$params)
{
  $this->thingExecutor->doTheThing(...$params);
}
```

In the latter case you just need to inject a different `thingExecutor` depending if you want to profile the thing or not.

Also, does it worth it adding this when you can use Blackfire.io?

**Comment by mickaelandrieu** [comment]
Hi @StanislovasRamanauskas,

would you mind to contribute this to the [Developer Tools](https://github.com/friends-of-presta/developer_tools) module ?

I will take this code anyway, but I prefer to let you the attribution of this code 👍, the way I see it this won't be shipped in 1.7.7 so we won't have this in the Core before 2021...


@eternoendless the main use of this profiler is * not * the performance, but figure out which hooks are available: this is a limited version of the Hook Profiler I've contributed on the "modern" part of PrestaShop.

EDIT (13/05/20) => Now available under [Friends of Presta Developer Tools module](https://github.com/friends-of-presta/developer_tools/pull/12)

**Comment by PierreRambaud** [comment]
Closed in favor of #20673
Thanks for your work, it helps us a lot :wink:

**Pull request closed without merging**